### ISSUE TITLE
Bug: fields annotated as `Any` are not required in v1.1

### ISSUE THREAD

**Issue by dmontagu** [samuelcolvin/pydantic#1028]
# Bug

In `v1.0`, the following code raises an error:

```py
from typing import Any

from pydantic import BaseModel


class A(BaseModel):
    x: Any

print(A())
"""
pydantic.error_wrappers.ValidationError: 1 validation error for A
x
  field required (type=value_error.missing)
"""
```

However, in v1.1, it fills the value with `None`:
```python
from typing import Any

from pydantic import BaseModel


class A(BaseModel):
    x: Any

print(A())
# x=None
```

--------

I personally prefer the v1.0 behavior, (not to mention the issue of api stability now that we are on v1). Note this causes problems for fastapi as noted in this comment: https://github.com/tiangolo/fastapi/pull/646#issuecomment-557943292 .

--------

I think the issue is the following code in `ModelField._type_analysis`:

```python
        if self.type_ is Any:
            self.required = False
            self.allow_none = True
            return
```
This makes the field non-required if it is of type `Any`; this definitely seems undesirable to me (setting `allow_none=True` obviously makes sense though).

@samuelcolvin is there a good reason to be setting `self.required = False` when annotating as `Any`? Is this required elsewhere? (Was this change to the behavior of fields annotated as `Any` intentional?)

**Comment by samuelcolvin** [comment]
This was done in #962 to fix #961.

I'm very sorry this has caused problems with fastapi, it was a fix for a bug. I hoped it would only be backwards incompatible insofar as it fixed the bug, I guess I was wrong.

The first problem is that pydantic until now has a policy of "if the field can be `None`, it is also omissible" - therefore since an `Any` field can be set to `None` it follows it should also be valid to omit that value. 

_(I'm using "omissible" not "optional" to avoid the confusion over the meaning of "optional" discussed in #990.)_

I'm open to changing this in v2, see #990 for a discussion of this, but obviously that can't be done before.

I don't think it makes sense for `Any` to have different behaviour for `None` compared to say `Union[None, str, int]`.

The second problem would be reversing the change introduced in v1.1 via #962 in v1.2. Surely one subtle change in a minor version is better than 2?

**Comment by dmontagu** [comment]
Yeah, I remembered there was a strange issue where this came up, couldn't remember it well enough to find it though, thanks for linking.

-----

I think your line of reasoning makes sense here justifying this behavior. But I think there should at least be a way to mark a field as required if it is annotated as type `Any`. As I proposed in the linked comment, currently just setting `field.required = True` seems to work for this.

For FastAPI's sake, I think one of the following three things should be done:
1. Treat arguments annotated/inferred as `Any` as optional *in FastAPI*
2. Add a way to explicitly mark a `ModelField` as required in a way that won't be overridden during type analysis, so that FastAPI can do this for non-`Optional` `Any` fields.
3. Confirm that setting `field.required = True` after the `__init__` call is the *intended* way to accomplish this.

Obviously option 3 requires the least work, so if you think that's what we should do in FastAPI I think that's fine.

Option 1 is the second least work, but I don't know what opinion @tiangolo might have on this point. I personally would find it slightly unintuitive given how FastAPI worked up to this point, but at the same time I almost never leave endpoint arguments un-annotated or annotated as `Any`, and I think this *could* be a totally reasonable convention for FastAPi (i.e., injecting `None` in the absence of an annotation). So I think it's up to @tiangolo.

I don't know how much work option 2 would be, or if it would be difficult to get right without breaking something else. It doesn't seem like it would be too bad, but I haven't looked deeply into it and could imagine challenges.

**Comment by samuelcolvin** [comment]
1. My immediate thought was that option 1 was preferable, but I can see it would be annoying if the behaviour changed now, then again upon upgrading to pydantic v2.

2. `RequiredOptional` as suggested in #990 might be a solution to option 2? But I imagine all that will be doing is setting `field.required = True`

3. I can't imagine a problem with this but it's the "unknown unknowns" that might be our downfall. I'll have a look through and add some tests to try and confirm that setting `field.required = True` after a field is initialised and prepared should work properly.

**Comment by tiangolo** [comment]
Yeah, I agree that if `Any` includes `None`, it would make sense to make a field annotated with `Any` as optional.

The problem is in a case like this:

```Python
from pydantic import BaseModel, Field

class A(BaseModel):
    x: Field(...)

print(A())
```

That, in theory, would mark the field as an implicit `Any`, but required.

In Pydantic 1.1 that's raising an error:

```
Traceback (most recent call last):
  File "pydantic/validators.py", line 557, in pydantic.validators.find_validators
TypeError: issubclass() arg 1 must be a class

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "sand.py", line 4, in <module>
    class A(BaseModel):
  File "pydantic/main.py", line 192, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 260, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 237, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 302, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 412, in pydantic.fields.ModelField._populate_validators
  File "pydantic/validators.py", line 566, in find_validators
RuntimeError: error checking inheritance of FieldInfo(default=Ellipsis, extra={}) (type: FieldInfo)
```

But in FastAPI, that syntax (without type hint) is allowed. That's where the breaking change is happening on the FastAPI side.

Here's the code that is breaking:

```Python
@app.get("/query/param-required")
def get_query_param_required(query=Query(...)):
    return f"foo bar {query}"
```

...I'm not really sure why the error is not thrown in FastAPI, I think I check it differently, not requiring a type declaration.

---

Now, the problem with option 1, with respect to FastAPI, is that this situation would be quite unintuitive:

For a FastAPI *path operation function* (a route handler) with a function like:

```Python
@app.get("/items/")
def get_items_for_user(username):
    return username
```

In this example (not a good API design, but valid nonetheless), a *Query* parameter of `username` would be expected.

The implicit type of `username` would be `Any`, so it would appear as optional.

But in a function in normal Python, a parameter without a default is a required parameter...

**Comment by samuelcolvin** [comment]
I see that, would option 2 or 3 work for fastapi?

**Comment by tiangolo** [comment]
You can disregard my previous comment, with respect to FastAPI, it's all solved with #1031.

I'm fine with option 1, treating `Any` as `Optional`. It's probably the most pythonic and "correct".

And for those edge cases above, FastAPI is already doing the checks and handling it, it should all work as expected with the PR above.

In fact, it's kind of Option 3, just that #1031 makes the change in option 3 unnecessary by handling it in Pydantic directly, making `FieldModel(required=True)` keep the `required` flag.
